scripts/upgrade-pwa.js

```diff
@@ -17,34 +21,75 @@ const parseVersion = version => {
   return version;
 };
 
-const modifyPackageJson = (cb, message) => {
-  const fs = require('fs');
+const writeVersion = version => {
+  if (/^[0-9]+$/.test(version)) {
+    return `^${version}.0.0`;
+  } else if (/^[0-9]+\.[0-9]+$/.test(version)) {
+    return `~${version}.0`;
+  } else {
+    return version;
+  }
+};
+
+const execute = command => {
+  console.log(`## ${command}`);
+  execSync(command, { stdio: 'inherit' });
+};
+
+const modifyPackageJson = cb => {
   const { parse, stringify } = require('comment-json');
   const packageJson = cb(parse(fs.readFileSync('./package.json', { encoding: 'UTF-8' })));
-  fs.writeFileSync('./package.json', stringify(packageJson, null, 2));
-  if (message) execSync('git commit -a -m "' + message + '"', { stdio: 'inherit' });
+  fs.writeFileSync('./package.json', stringify(packageJson, null, 2) + '\n');
+  execute('npx prettier --write package.json');
 };
 
+const commit = message => {
+  execute(`git commit -a --allow-empty -m "${message}"`, { stdio: 'inherit' });
+};
+
+const angularCoreDependencies = ['zone.js', 'tslib', 'rxjs', 'typescript'];
+
+const NPM = process.platform === 'win32' ? 'npm.cmd' : 'npm';
+
+// </HELPERS>
+
 // synchronize node_modules in working copy
-execSync('npm ci', { stdio: 'inherit' });
+execute('npm ci');
 
 // deactivate full postinstall
 let install;
 modifyPackageJson(packageJson => {
   install = packageJson.scripts.postinstall;
   packageJson.scripts.postinstall = 'postinstall';
   return packageJson;
-}, 'chore: temporarily deactivate full postinstall');
+});
+commit('chore: temporarily deactivate full postinstall');
 
 // upgrade core libraries
 console.log('upgrade @schematics/angular to find compatible versions');
-execSync('npx ng update @schematics/angular', { stdio: 'inherit' });
+execute('npx ng update @schematics/angular');
 
 const { latestVersions } = require('@schematics/angular/utility/latest-versions');
-pinned['zone.js'] = parseVersion(latestVersions['zone.js']);
-pinned.typescript = parseVersion(latestVersions.typescript);
-pinned.tslib = parseVersion(latestVersions.tslib);
-pinned.rxjs = parseVersion(latestVersions.rxjs);
+angularCoreDependencies.forEach(dep => {
+  pinned[dep] = parseVersion(latestVersions[dep]);
+});
+
+// set pinned version in package.json
+modifyPackageJson(packageJson => {
+  Object.entries(pinned).forEach(([dep, ver]) => {
+    const write = writeVersion(ver);
+    if (packageJson.dependencies[dep]) {
+      packageJson.dependencies[dep] = write;
+    } else {
+      packageJson.devDependencies[dep] = write;
+    }
+  });
+  return packageJson;
+});
+
+// install pinned versions
+execute('npm i');
+commit('chore: install pinned versions');
```

Comment:
Why is this step needed that temporary installs older versions of the pinned dependencies?

![image](https://user-images.githubusercontent.com/50741106/136775538-7773505d-fb96-4bbb-8478-2ffbe80bc5d0.png)

Reply:
It doesn't install "older" dependencies. `^4.0.0` means for example: take the newest major 4 version. `~4.0.0` would mean: take the newest 4.0 minor version, and so on.
See also: https://docs.npmjs.com/about-semantic-versioning#using-semantic-versioning-to-specify-update-types-your-package-can-accept

In this temporary step, the right pinned versions are just set in the package.json, so they are respected in the next step.
Later on there is a step that puts in the actual versions.